pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java

```diff
@@ -356,14 +355,15 @@ public void enableCnxAutoRead() {
     }
 
     protected boolean hasLocalProducers() {
-        AtomicBoolean foundLocal = new AtomicBoolean(false);
-        producers.values().forEach(producer -> {
+        if (producers.isEmpty()) {
+            return false;
+        }
+        for (Producer producer : producers.values()) {
```

Comment:
`producers.values()` is taking a copy of the map so it would be very expensive.

Reply:
It's create a reusable `ValuesView` instance under `ConcurrentHashMap#values`, not a copy I think.

Reply:
It's create a reusable ValuesView instance under ConcurrentHashMap#values, not a copy I think.

Reply:
Oh, you're right, we using the regular `ConcurrentHashMap` here instead of the other custom map.